File: RPC/RPC_10_12_10_2024/Soluciones/ThirteenthAnnual_UCF_LocalProgrammingContest/local2024FinalRoundData/solutions/K_teamwork/teamwork.cpp

#include <bits/stdc++.h>

#define sz(a) ((int)(a).size())
#define rep(a,b,c) for(int a = (b); a < (c); a++)

using namespace std;
using vi = vector<int>;
using vvi = vector<vi>;

// Thank you kactl
pair<int, vi> hungarian(const vector<vi> &a) {
	if (a.empty()) return {0, {}};
	int n = sz(a) + 1, m = sz(a[0]) + 1;
	vi u(n), v(m), p(m), ans(n - 1);
	rep(i,1,n) {
		p[0] = i;
		int j0 = 0; // add "dummy" worker 0
		vi dist(m, INT_MAX), pre(m, -1);
		vector<bool> done(m + 1);
		do { // dijkstra
			done[j0] = true;
			int i0 = p[j0], j1, delta = INT_MAX;
			rep(j,1,m) if (!done[j]) {
				auto cur = a[i0 - 1][j - 1] - u[i0] - v[j];
				if (cur < dist[j]) dist[j] = cur, pre[j] = j0;
				if (dist[j] < delta) delta = dist[j], j1 = j;
			}
			rep(j,0,m) {
				if (done[j]) u[p[j]] += delta, v[j] -= delta;
				else dist[j] -= delta;
			}
			j0 = j1;
		} while (p[j0]);
		while (j0) { // update alternating path
			int j1 = pre[j0];
			p[j0] = p[j1], j0 = j1;
		}
	}
	rep(j,1,m) if (p[j]) ans[p[j] - 1] = j - 1;
	return {-v[0], ans}; // min cost
}

int canDo(vvi & orig, int lim, int ps) {
    int n = sz(orig), m = sz(orig[0]);
    vvi newTab(n, vi(m + (n-ps), 0));
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < m; j++) { 
            newTab[i][j] = orig[i][j];
        }
    }
    return (hungarian(newTab).first <= lim);
}

int main() {
    int stud, prob, time_lim, pairs;
    cin >> stud >> prob >> time_lim;
    vvi orig(stud, vi(prob, time_lim + 1));
    //for (int i = 0; i < pairs; i++) {
    for (int i = 0; i < stud; i++) {
        for (int j = 0; j < prob; j++) {
            int sid, pid, tim;
            cin >> tim;
            sid--; pid--;
            sid = i;
            pid = j;
            if (!tim) tim = time_lim + 1;
            orig[sid][pid] = min(orig[sid][pid], tim);
        }
    }

    int lo = 1, hi = min(stud, prob);
    int ans = 0;
    while (lo <= hi) {
        int mid = (lo + hi) / 2;
        if (canDo(orig, time_lim, mid)) {
            ans = mid; lo = mid + 1;
        } else {
            hi = mid - 1;
        }
    }
    cout << ans << endl;
    return 0;
}

File: RPC/RPC_10_12_10_2024/Soluciones/ThirteenthAnnual_UCF_LocalProgrammingContest/local2024FinalRoundData/solutions/K_teamwork/teamwork.java
// Arup Guha
// 8/26/2024
// Solution to 2024 UCF Locals Problem: Teamwork

import java.util.*;

public class teamwork {

	// For Hungarian...
	public static int[] rowMatchesWith;

	// Original data is stored here.
	public static int[][] mat;
	public static int nStud;
	public static int nProb;
	public static int limit;
	
	public static void main(String[] args) {
	
		// Read basic data.
		Scanner stdin = new Scanner(System.in);
		nStud = stdin.nextInt();
		nProb = stdin.nextInt();
		limit = stdin.nextInt();
		
		// Hungarian matching matrix. Just stores original values.
		mat = new int[nProb][nStud];
		
		// Read in times. If you can't do a problem, we make it really big.
		for (int j=0; j<nStud; j++) {
			for (int i=0; i<nProb; i++) { 
				mat[i][j] = stdin.nextInt();
				if (mat[i][j] == 0)
					mat[i][j] = 1000000;
			}
		}
		
		// Limits for binary search.
		int low = 0, high = Math.min(nProb, nStud);
		
		// Run a binary search.
		while (low < high) {
			
			// # of problems we are trying to solve.
			int mid = (low+high+1)/2;
			
			// We can do it in time, answer is atleast mid.
			if (canDo(mid))
				low = mid;
			
			// We can't do mid, so mid-1 is most we could possibly do.
			else
				high = mid-1;
		}
		
		// Ta da!
		System.out.println(low);
	}
	
	// Returns true iff we can solve solved # of problems in time.
	public static boolean canDo(int solved) {
		
		// These are the unsolved problems.
		int unsolved = nProb - solved;
		
		// So we create some perfect students, who will solve the unsolved problems in 0 time.
		int[][] m = new int[nProb][nStud+unsolved];
		
		// Just copy the original values, and the 0s are for those perfect students.
		for (int j=0; j<nStud; j++) 
			for (int i=0; i<nProb; i++) 
				m[i][j] = mat[i][j];
				
		// Run Hungarian; if total cost is less than limit, we can solve solved # of problems.
		int mincost = hungarian(m);
		return mincost <= limit;
	}
	
	/*** UCF Hackpack for Hungarian circa Sage, Kamino era. ***/
	public static int hungarian(int[][] a) {

		int n=a.length, m=a[0].length; if (n>m) throw null;
		int[] u=new int[n], v=new int[m+1]; //edge (i->j) >= u[i]+v[j]
		int[] p=new int[m+1]; //p[j] = left match for right node j
		Arrays.fill(p, -1);

		for (int i=0; i<n; i++) {//find alternating path for node i
			int j0=m; p[j0]=i;
			int[] dist=new int[m+1], from=new int[m+1];
			boolean[] seen=new boolean[m+1];
			Arrays.fill(dist, Integer.MAX_VALUE);
			Arrays.fill(from, -1);

			while (p[j0]!=-1) {

				seen[j0]=true; int i0=p[j0], delta=Integer.MAX_VALUE, j1=-1;
				for (int j=0; j<m; j++) {//consider edges i0 -> everything
					if (seen[j]) continue;
					int candDist=a[i0][j]-u[i0]-v[j];
					if (candDist<dist[j]) {dist[j]=candDist; from[j]=j0;}
					if (dist[j]<delta) {delta=dist[j]; j1=j;}
				}
				
				//it costs at least delta to get somewhere else,
				//subtract that from all distances and add cost to u, v, arrays
				//from all done -> not done edges
				for (int j=0; j<=m; j++) {
					if (seen[j]) {u[p[j]]+=delta; v[j]-=delta;}
					else dist[j]-=delta;
				}

				j0=j1;
			}
			
			//flip alternating path
			while (j0!=m) { int j1=from[j0]; p[j0]=p[j1]; j0=j1; }
		}
		
		//sum of deltas is stored at v[m] coincidentally
		int ans=-v[m];
		rowMatchesWith=new int[n];
		for (int j=0; j<m; j++) 
			if (p[j]!=-1) rowMatchesWith[p[j]]=j;

		return ans;
	}

}
